File: apps/web/src/lib/simulation/parse-addon-export.test.ts
import { describe, expect, it } from "vitest";
import { EquipmentSlot, EquipmentSource } from "@/lib/saint-api/generated";
import {
	normalizeLineEndings,
	parseEquipmentItem,
	parseSimcAddonExport,
} from "./parse-addon-export";

describe("parseSimcAddonExport", () => {
	it("parses character metadata and improves on Go by preserving character name", () => {
		const got = parseSimcAddonExport(
			'priest="Example"\nlevel=80\nrace=void_elf\nregion=us\nserver=area-52\nrole=spell\nprofessions=alchemy=100/herbalism=100\nspec=shadow',
		);

		expect(got).toMatchObject({
			name: "Example",
			character_class: "priest",
			level: 80,
			race: "void_elf",
			region: "us",
			server: "area-52",
			role: "spell",
			professions: "alchemy=100/herbalism=100",
			spec: "shadow",
			equipped_items: [],
		});
	});

	it("splits active talents from non-active saved loadouts", () => {
		const got = parseSimcAddonExport(`# Saved Loadout: Dungeon
# talents=DUNGEON_TALENTS
talents=ACTIVE_TALENTS`);

		expect(got.active_talents).toEqual({
			name: "Active",
			talents: "ACTIVE_TALENTS",
		});
		expect(got.talent_loadouts).toEqual([
			{
				name: "Dungeon",
				talents: "DUNGEON_TALENTS",
			},
		]);
	});

	it("parses equipped and bag item groups from addon export sections", () => {
		const got = parseSimcAddonExport(`# Host Commander's Casque (684)
head=,id=250458,bonus_id=6652/12667,gem_id=240865,ilevel=684
# Gear from Bags
# Gnarlroot Spinecleaver (710)
main_hand=,id=249671,enchant_id=3368,bonus_id=6652,ilevel=710`);

		expect(got.equipped_items).toHaveLength(1);
		expect(got.equipped_items[0]).toMatchObject({
			slot: EquipmentSlot.HEAD,
			name: "Host Commander's Casque",
			display_name: "Host Commander's Casque",
			item_id: 250458,
			item_level: 684,
			bonus_ids: [6652, 12667],
			gem_ids: [240865],
			source: EquipmentSource.EQUIPPED,
			raw_line: "head=,id=250458,bonus_id=6652/12667,gem_id=240865,ilevel=684",
		});

		expect(got.bag_items).toHaveLength(1);
		expect(got.bag_items?.[0]).toMatchObject({
			slot: EquipmentSlot.MAIN_HAND,
			name: "Gnarlroot Spinecleaver",
			item_id: 249671,
			item_level: 710,
			enchant_id: 3368,
			source: EquipmentSource.BAG,
		});
	});

	it("parses Additional Character Info comments", () => {
		const got = parseSimcAddonExport(`# loot_spec=shadow
# Additional Character Info
# catalyst_currencies=3116:2/2912:1
# slot_high_watermarks=head:684:710/main_hand:710:720
# upgrade_achievements=1/2`);

		expect(got.loot_spec).toBe("shadow");
		expect(got.catalyst_currencies).toEqual([
			{ id: 2912, quantity: 1 },
			{ id: 3116, quantity: 2 },
		]);
		expect(got.slot_high_watermarks).toEqual([
			{
				slot: EquipmentSlot.HEAD,
				current_item_level: 684,
				max_item_level: 710,
			},
			{
				slot: EquipmentSlot.MAIN_HAND,
				current_item_level: 710,
				max_item_level: 720,
			},
		]);
		expect(got.upgrade_achievements).toEqual([1, 2]);
	});

	it("normalizes line endings before parsing", () => {
		expect(
			normalizeLineEndings('priest="Example"\r\nlevel=80\rspec=shadow'),
		).toBe('priest="Example"\nlevel=80\nspec=shadow');

		const got = parseSimcAddonExport(
			'priest="Example"\r\nlevel=80\rspec=shadow',
		);

		expect(got.level).toBe(80);
		expect(got.spec).toBe("shadow");
	});
});

describe("parseEquipmentItem", () => {
	it("parses optional equipment attributes and keeps raw line intact", () => {
		const rawLine =
			"head=,id=250458,enchant_id=3368,crafting_quality=5,bonus_id=6652/12667,gem_id=240865,gem_id2=240866,crafted_stats=36/40,drop_level=684";

		const got = parseEquipmentItem(
			"Host Commander's Casque (700)",
			rawLine,
			EquipmentSource.EQUIPPED,
		);

		expect(got).toMatchObject({
			slot: EquipmentSlot.HEAD,
			name: "Host Commander's Casque",
			display_name: "Host Commander's Casque",
			item_id: 250458,
			item_level: 700,
			enchant_id: 3368,
			crafting_quality: 5,
			bonus_ids: [6652, 12667],
			gem_ids: [240865, 240866],
			crafted_stats: [36, 40],
			source: EquipmentSource.EQUIPPED,
			raw_line: rawLine,
		});
	});

	it("returns undefined for unrecognized slots or missing item ids", () => {
		expect(
			parseEquipmentItem("", "unknown=,id=1", EquipmentSource.EQUIPPED),
		).toBeUndefined();
		expect(
			parseEquipmentItem("", "head=,bonus_id=6652", EquipmentSource.EQUIPPED),
		).toBeUndefined();
	});
});

File: apps/api/api_utils/parse_addon_export_test.go
//nolint:exhaustruct
package api_utils

import (
	"testing"

	api "github.com/DomNidy/saint_sim/internal/api"
)

func TestParseAddonExportToWowCharacter(t *testing.T) {
	t.Parallel()

	got := ParseAddonExport(`priest="Example"
level=80
race=void_elf
region=us
server=area-52
role=spell
professions=alchemy=100/herbalism=100
spec=shadow
talents=ACTIVE_TALENTS
# loot_spec=shadow
# Saved Loadout: Dungeon
# talents=DUNGEON_TALENTS
# Host Commander's Casque (684)
head=,id=250458,bonus_id=6652/12667,gem_id=240865,ilevel=684
# Gear from Bags
# Gnarlroot Spinecleaver (710)
main_hand=,id=249671,bonus_id=6652,ilevel=710
# Additional Character Info
# catalyst_currencies=2912:1/3116:2
# slot_high_watermarks=head:684:710/main_hand:710:720
# upgrade_achievements=1/2`)

	if got.CharacterClass != api.Priest {
		t.Fatalf("character_class = %q, want %q", got.CharacterClass, api.Priest)
	}
	if got.Level != 80 {
		t.Fatalf("level = %d, want 80", got.Level)
	}
	if got.Spec != "shadow" {
		t.Fatalf("spec = %q, want shadow", got.Spec)
	}
	if got.LootSpec == nil || *got.LootSpec != "shadow" {
		t.Fatalf("loot_spec = %v, want shadow", got.LootSpec)
	}
	if got.ActiveTalents == nil || got.ActiveTalents.Talents != "ACTIVE_TALENTS" {
		t.Fatalf("active_talents = %#v, want ACTIVE_TALENTS", got.ActiveTalents)
	}
	if got.TalentLoadouts == nil || len(*got.TalentLoadouts) != 1 {
		t.Fatalf("talent_loadouts = %#v, want one non-active loadout", got.TalentLoadouts)
	}
	if (*got.TalentLoadouts)[0].Name == nil || *(*got.TalentLoadouts)[0].Name != "Dungeon" {
		t.Fatalf("talent loadout name = %#v, want Dungeon", (*got.TalentLoadouts)[0].Name)
	}
	if (*got.TalentLoadouts)[0].Talents != "DUNGEON_TALENTS" {
		t.Fatalf(
			"talent loadout talents = %q, want DUNGEON_TALENTS",
			(*got.TalentLoadouts)[0].Talents,
		)
	}
	if len(got.EquippedItems) != 1 {
		t.Fatalf("equipped_items len = %d, want 1", len(got.EquippedItems))
	}
	if got.EquippedItems[0].Source != api.Equipped {
		t.Fatalf("equipped item source = %q, want %q", got.EquippedItems[0].Source, api.Equipped)
	}
	if got.BagItems == nil || len(*got.BagItems) != 1 {
		t.Fatalf("bag_items = %#v, want one bag item", got.BagItems)
	}
	if (*got.BagItems)[0].Source != api.Bag {
		t.Fatalf("bag item source = %q, want %q", (*got.BagItems)[0].Source, api.Bag)
	}
	if got.CatalystCurrencies == nil || len(*got.CatalystCurrencies) != 2 {
		t.Fatalf("catalyst_currencies = %#v, want two entries", got.CatalystCurrencies)
	}
	if (*got.CatalystCurrencies)[0].Id != 2912 || (*got.CatalystCurrencies)[0].Quantity != 1 {
		t.Fatalf("first catalyst currency = %#v, want 2912:1", (*got.CatalystCurrencies)[0])
	}
	if got.SlotHighWatermarks == nil || len(*got.SlotHighWatermarks) != 2 {
		t.Fatalf("slot_high_watermarks = %#v, want two entries", got.SlotHighWatermarks)
	}
	if (*got.SlotHighWatermarks)[0].Slot != api.Head {
		t.Fatalf(
			"first slot watermark slot = %q, want %q",
			(*got.SlotHighWatermarks)[0].Slot,
			api.Head,
		)
	}
	if got.UpgradeAchievements == nil || len(*got.UpgradeAchievements) != 2 {
		t.Fatalf("upgrade_achievements = %#v, want two entries", got.UpgradeAchievements)
	}
}

func TestNormalizeLineEndings(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name  string
		input string
		want  string
	}{
		{
			name:  "already normalized",
			input: "line1\nline2\nline3",
			want:  "line1\nline2\nline3",
		},
		{
			name:  "windows line endings",
			input: "line1\r\nline2\r\nline3",
			want:  "line1\nline2\nline3",
		},
		{
			name:  "classic mac line endings",
			input: "line1\rline2\rline3",
			want:  "line1\nline2\nline3",
		},
		{
			name:  "mixed line endings",
			input: "line1\r\nline2\rline3\nline4",
			want:  "line1\nline2\nline3\nline4",
		},
		{
			name:  "empty string",
			input: "",
			want:  "",
		},
		{
			name:  "single trailing carriage return",
			input: "line1\r",
			want:  "line1\n",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			got := NormalizeLineEndings(tt.input)
			if got != tt.want {
				t.Fatalf("NormalizeLineEndings(%q) = %q, want %q", tt.input, got, tt.want)
			}
		})
	}
}

func TestStripAllComments(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name  string
		input string
		want  string
	}{
		{
			name:  "Removes only commented lines",
			input: "# Simc version 1.23\nwarrior=John\nlevel=90\n",
			want:  "warrior=John\nlevel=90\n",
		},
		{
			name:  "Only commented input returns empty str",
			input: "# This is a comment line1.\n#Comment line 2\n#asd\n",
			want:  "",
		},
		{
			name:  "Multiple consecutive #'s",
			input: "##Comment #Two\n## Hey\nwarrior=John\n###123",
			want:  "warrior=John",
		},
		{
			name:  "Preserves blank lines between non-comment lines",
			input: "# Simc version 1.23\n\nwarrior=John\n\nlevel=90\n# trailing comment",
			want:  "\nwarrior=John\n\nlevel=90",
		},
		{
			name:  "Preserves spacing on non-comment lines",
			input: "# comment\n  warrior=John  \n\tlevel=90\t\n",
			want:  "  warrior=John  \n\tlevel=90\t\n",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			got := StripAllComments(tt.input)
			if got != tt.want {
				t.Fatalf("StripAllComments(%q) = %q, want %q", tt.input, got, tt.want)
			}
		})
	}
}

func TestTrimLineWhitespace(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name  string
		input string
		want  string
	}{
		{
			name:  "Trims leading and trailing spaces on each line",
			input: "  warrior=John  \n  level=90  ",
			want:  "warrior=John\nlevel=90",
		},
		{
			name:  "Preserves blank lines while trimming spaces",
			input: "  warrior=John  \n   \n  level=90  ",
			want:  "warrior=John\n\nlevel=90",
		},
		{
			name:  "Normalizes line endings before trimming",
			input: "  warrior=John  \r\n  level=90  \r",
			want:  "warrior=John\nlevel=90\n",
		},
		{
			name:  "Trims tabs at the start and end of each line",
			input: "\twarrior=John\t\n \tlevel=90\t ",
			want:  "warrior=John\nlevel=90",
		},
		{
			name:  "Preserves internal tabs",
			input: "value\t=\t42\n\tname\t=\tvalue\t",
			want:  "value\t=\t42\nname\t=\tvalue",
		},
		{
			name:  "Empty string",
			input: "",
			want:  "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			got := TrimLineWhitespace(tt.input)
			if got != tt.want {
				t.Fatalf("TrimLineWhitespace(%q) = %q, want %q", tt.input, got, tt.want)
			}
		})
	}
}
